packages/typespec-client-generator-core/src/internal-utils.ts

```diff
@@ -252,12 +255,85 @@ export function getSdkTypeBaseHelper<TKind>(
   context: TCGCContext,
   type: Type,
   kind: TKind
-): DefaultSdkTypeBase<TKind> {
-  return {
+): [DefaultSdkTypeBase<TKind>, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  return diagnostics.wrap({
     __raw: type,
     deprecation: getDeprecationDetails(context.program, type)?.message,
     kind,
-  };
+    decorators: diagnostics.pipe(getTypeDecorators(context, type)),
+  });
+}
+
+export function getTypeDecorators(
+  context: TCGCContext,
+  type: Type
+): [Record<string, Record<string, any>>, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  const retval: Record<string, Record<string, any>> = {};
+  if ("decorators" in type) {
+    for (const decorator of type.decorators) {
+      // only process explicitly defined decorators
+      if (decorator.definition) {
+        const decoratorName = `${decorator.definition?.namespace ? getNamespaceFullName(decorator.definition.namespace) + "." : ""}${decorator.definition?.name}`;
+        // white list filtering
+        if (
+          !context.decoratorsWhiteList ||
+          !context.decoratorsWhiteList.some(
+            (x) => x === decoratorName || new RegExp(x).test(decoratorName)
+          )
+        ) {
+          continue;
+        }
+
+        retval[decoratorName] = {};
+        for (let i = 0; i < decorator.definition.parameters.length; i++) {
+          if (i >= decorator.args.length) {
+            retval[decoratorName][decorator.definition.parameters[i].name] = undefined;
```

Comment:
Is it possible to skip nullable parameter by using named parameter in typespec decorator?

Reply:
no

Reply:
Then, can we skip the undefined arguments?
We can direct loop through `decorator.args` instead of `decorator.definition.parameters`, and there is pointless to store the undefined arguments.

Reply:
i think keep all the decorator info is much better, what do you think?

Reply:
Not really, the arguments are not defined in typespec anyway, they will not be used in emitter.